### PR TITLE
[APM-128] Search offsets

### PR DESCRIPTION
**Pull request by gnusi**
### Scope & Purpose

Community part of APM-128.

- [ ] :hankey: Bugfix
- [x] :pizza: New feature
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [ ] **Regression tests**
  - [ ] C++ **Unit tests**
  - [ ] **integration tests**
  - [ ] **resilience tests**
- [ ] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [ ] Backports
  - [ ] Backport for 3.9: *(Please link PR)*
  - [ ] Backport for 3.8: *(Please link PR)*
  - [ ] Backport for 3.7: *(Please link PR)*

#### Related Information

*(Please reference tickets / specification / other PRs etc)*

- [ ] Docs PR: 
- [x] Enterprise PR: https://github.com/arangodb/enterprise/pull/979
- [ ] GitHub issue / Jira ticket:
- [ ] Design document:

### CODE REVIEW

File: arangod/IResearch/IResearchFeature.cpp [hunk 1]
```diff
@@ -21,6 +21,9 @@
 /// @author Andrey Abramov
 /// @author Vasily Nabatchikov
 ////////////////////////////////////////////////////////////////////////////////
+
+#include "IResearchFeature.h"
```
Comment (MBkkt) [inline arangod/IResearch/IResearchFeature.cpp:25]:
```suggestion
#include "IResearch/IResearchFeature.h"
```

File: arangod/IResearch/IResearchOrderFactory.cpp [hunk 2]
```diff
@@ -39,54 +41,17 @@
...
-
-    if (!arg || !arg->isDeterministic()) {
-      // we don't support non-deterministic arguments for scorers
-      return nullptr;
-    }
-  }
-
-  return reinterpret_cast<arangodb::aql::Variable const*>(arg0->getData());
-}
+using namespace arangodb;
```
Comment (MBkkt) [inline arangod/IResearch/IResearchOrderFactory.cpp:51]: why don't u put anon namespace to arangodb, that written bellow?

File: arangod/IResearch/SearchFunc.h [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+////////////////////////////////////////////////////////////////////////////////
+
+#include "Aql/Ast.h"
+#include "Aql/AstNode.h"
+#include "Aql/ExecutionNode.h"
+#include "Aql/Expression.h"
+#include "Containers/FlatHashMap.h"
+#include "IResearch/AqlHelper.h"
+
+#pragma once
```
Comment (MBkkt) [inline arangod/IResearch/SearchFunc.h:31]: before include? should be in codestyle?

File: arangod/IResearch/SearchFunc.h [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+                       aql::AstNode const* node) noexcept
+      : var{var}, node{node} {}
+
+  constexpr bool operator==(SearchFunc const& rhs) const noexcept {
+    return var == rhs.var && node == rhs.node;
+  }
+
+  constexpr bool operator!=(SearchFunc const& rhs) const noexcept {
+    return !(*this == rhs);
+  }
```
Comment (MBkkt) [inline arangod/IResearch/SearchFunc.h:49]: is it necessary?
I thought from c++20 enough == + <=>

File: arangod/IResearch/SearchFunc.h [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+  aql::Variable const* var{};  // scorer variable
+  aql::AstNode const* node{};  // scorer node
+};
+
+struct HashedSearchFunc : SearchFunc {
+  HashedSearchFunc(aql::Variable const* var, aql::AstNode const* node)
+      : SearchFunc{var, node}, hash{arangodb::iresearch::hash(node)} {}
```
Comment (MBkkt) [inline arangod/IResearch/SearchFunc.h:57]: equal or hash can throw exception?

File: arangod/IResearch/SearchFuncReplace.cpp [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+      for (auto* v : usedVars) {
+        if (!validVars.contains(v)) {
+          TRI_ASSERT(func.node);
+          auto const funcName = iresearch::getFuncName(*func.node);
+
+          THROW_ARANGO_EXCEPTION_FORMAT(
+              TRI_ERROR_BAD_PARAMETER,
+              "Inaccesible non-ArangoSearch view variable '%s' is used in "
+              "search function '%s'",
+              v->name.c_str(), funcName.c_str());
```
Comment (MBkkt) [inline arangod/IResearch/SearchFuncReplace.cpp:55]: Now we have three option for that 😢, and I think two of them better than this
`THROW_ARANGO_EXCEPTION_MESSAGE` +
1) absl::Substitute("....$0...$1...", v->name, funcName) or absl::StrCat
2) fmt::format, also we have static fmt function for Result, and if I remember correct we can throw result

File: arangod/IResearch/SearchFuncReplace.cpp [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+#include "Aql/Ast.h"
+#include "Aql/IResearchViewNode.h"
+
+namespace arangodb::iresearch {
+
+void extractSearchFunc(IResearchViewNode const& viewNode,
+                       DedupSearchFuncs& dedup,
+                       std::vector<SearchFunc>& funcs) {
+  auto* viewVar = &viewNode.outVariable();
+  arangodb::aql::VarSet usedVars;
```
Comment (MBkkt) [inline arangod/IResearch/SearchFuncReplace.cpp:35]:
```suggestion
  aql::VarSet usedVars;
```

File: arangod/IResearch/SearchFuncReplace.h [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+
+  if (exprNode != newNode) {
+    // simple expression, e.g LET x = BM25(d)
+    expr.replaceNode(newNode);
+  } else if (const bool hasFunc = !arangodb::iresearch::visit<true>(
+                 *exprNode,
+                 [&refExtractor](aql::AstNode const& node) {
+                   return nullptr == refExtractor(node);
+                 });
+             hasFunc) {
```
Comment (MBkkt) [inline arangod/IResearch/SearchFuncReplace.h:86]: don't like it, maybe early return?
```C++
  return;
}
const bool hasFunc = ...;
if (!hasFunc/hasFunc)

**Review by MBkkt — APPROVED** [review]

File: arangod/Aql/IResearchViewExecutor.h [hunk 9]
```diff
@@ -516,11 +536,12 @@ class IResearchViewExecutorBase {
   FilterCtx _filterCtx;  // filter context
   iresearch::ViewSnapshotPtr _reader;
   irs::filter::prepared::ptr _filter;
+  irs::filter::prepared const** _filterCookie{};
   irs::Order _order;
-  std::vector<ColumnIterator>
-      _storedValuesReaders;  // current stored values readers
+  std::vector<ColumnIterator> _storedValuesReaders;
+  std::array<char, arangodb::iresearch::kSearchDocBufSize> _buf;
```
Comment (MBkkt) [inline arangod/Aql/IResearchViewExecutor.h:542]:
```suggestion
  std::array<char, iresearch::kSearchDocBufSize> _buf;
```

File: arangod/IResearch/IResearchAnalyzerFeature.cpp [hunk 10]
```diff
@@ -2997,23 +2997,26 @@ void IResearchAnalyzerFeature::cleanupAnalyzers(irs::string_ref database) {
...
   WRITE_LOCKER(lock, _mutex);
   auto database = irs::string_ref(vocbase.name());
-  auto itr = _lastLoad.find(
-      static_cast<std::string>(  // FIXME: after C++20 remove cast and use
-                                 // heterogeneous lookup
-          irs::make_hashed_ref(database, std::hash<irs::string_ref>())));
+  // FIXME: after C++20 remove cast and use
+  // heterogeneous lookup
+  auto itr = _lastLoad.find(static_cast<std::string>(
+      irs::make_hashed_ref(database, std::hash<irs::string_ref>())));
```
Comment (MBkkt) [inline arangod/IResearch/IResearchAnalyzerFeature.cpp:3003]: Still doesn't work?
Maybe we should remove comment in that case?
Or use another map type

File: arangod/IResearch/IResearchKludge.cpp [hunk 11]
```diff
@@ -112,12 +112,35 @@ void mangleNested(std::string& name) {
...
+  if (isSearchFilter) {
     name += kAnalyzerDelimiter;
+
     name += analyzer._shortName;
   } else {
     mangleString(name);
+    if (analyzer->requireMangled()) {
+      name += kAnalyzerDelimiter;
+      name += analyzer._shortName;
+    }
```
Comment (MBkkt) [inline arangod/IResearch/IResearchKludge.cpp:124]: We discussed and decide didn't do it?
I also think if needed we can achieve it another way (We should have fieldAnalyzerProvider in 'exists', etc)

File: arangod/IResearch/SearchDoc.h [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+
+  irs::sub_reader const* segment() const noexcept { return _segment; }
+
+  irs::doc_id_t doc() const noexcept { return _doc; }
+
+  bool isValid() const noexcept {
+    return _segment && irs::doc_limits::valid(_doc);
+  }
+
+  constexpr auto operator<=>(const SearchDoc&) const noexcept = default;
```
Comment (MBkkt) [inline arangod/IResearch/SearchDoc.h:67]:
```suggestion
  constexpr auto operator==(SearchDoc const&) const noexcept = default;
  constexpr auto operator<=>(SearchDoc const&) const noexcept = default;
```

Spaceship is enough for comparison but not enough for effective comparison

File: arangod/IResearch/SearchDoc.h [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+
+  std::string_view encode(
+      std::span<char, kSearchDocBufSize> buf) const noexcept {
+    std::memcpy(buf.data(), static_cast<void const*>(&_segment),
+                sizeof _segment);
+    std::memcpy(buf.data() + sizeof _segment, &_doc, sizeof _doc);
```
Comment (MBkkt) [inline arangod/IResearch/SearchDoc.h:73]: I think we should make explicit cast in both case or none

File: arangod/IResearch/SearchDoc.h [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+}
+
+namespace arangodb::iresearch {
+
+constexpr size_t kSearchDocBufSize = sizeof(size_t) + sizeof(irs::doc_id_t);
+
+class SearchDoc {
+ public:
+  static SearchDoc decode(std::string_view buf) noexcept {
+    if (ADB_LIKELY(kSearchDocBufSize == buf.size())) {
```
Comment (MBkkt) [inline arangod/IResearch/SearchDoc.h:45]: NIT
early return?

File: tests/Aql/RegisterPlanTest.cpp [hunk 15]
```diff
@@ -209,7 +209,7 @@ class RegisterPlanTest : public ::testing::Test {
   auto walk(std::vector<ExecutionNodeMock>& nodes)
       -> std::shared_ptr<RegisterPlanT<ExecutionNodeMock>> {
     // Compute the variable usage for nodes.
-    std::unordered_map<VariableId, ExecutionNodeMock*> varSetBy;
+    arangodb::containers::FlatHashMap<VariableId, ExecutionNodeMock*> varSetBy;
```
Comment (MBkkt) [inline tests/Aql/RegisterPlanTest.cpp:212]:
```suggestion
    containers::FlatHashMap<VariableId, ExecutionNodeMock*> varSetBy;
```

File: tests/Aql/RegisterPlanTest.cpp [hunk 16]
```diff
@@ -329,7 +329,7 @@ class RegisterPlanTest : public ::testing::Test {
   }
 
   auto getVarUsage(std::vector<ExecutionNodeMock>& nodes) {
-    std::unordered_map<VariableId, ExecutionNodeMock*> varSetBy;
+    arangodb::containers::FlatHashMap<VariableId, ExecutionNodeMock*> varSetBy;
```
Comment (MBkkt) [inline tests/Aql/RegisterPlanTest.cpp:332]:
```suggestion
    containers::FlatHashMap<VariableId, ExecutionNodeMock*> varSetBy;
```

File: arangod/Aql/IResearchViewNode.h [hunk 17]
```diff
@@ -295,7 +308,7 @@ class IResearchViewNode final : public aql::ExecutionNode {
     bool canVariablesBeReplaced(aql::CalculationNode* calculationNode) const;
 
     ViewVarsInfo replaceViewVariables(
-        std::vector<aql::CalculationNode*> const& calcNodes,
+        std::span<aql::CalculationNode*> calcNodes,
```
Comment (MBkkt) [inline arangod/Aql/IResearchViewNode.h:311]:
```suggestion
        std::span<aql::CalculationNode* const> calcNodes,
```

File: arangod/Aql/IResearchViewNode.cpp [hunk 18]
```diff
@@ -1854,7 +1893,7 @@ void IResearchViewNode::OptimizationState::saveCalcNodesForViewVariables(
 
 IResearchViewNode::ViewVarsInfo
 IResearchViewNode::OptimizationState::replaceViewVariables(
-    std::vector<aql::CalculationNode*> const& calcNodes,
+    std::span<aql::CalculationNode*> calcNodes,
```
Comment (MBkkt) [inline arangod/Aql/IResearchViewNode.cpp:1896]:
```suggestion
    std::span<aql::CalculationNode* const> calcNodes,
```

File: arangod/IResearch/IResearchKludge.cpp [hunk 19]
```diff
@@ -114,10 +114,33 @@ void mangleField(std::string& name, bool isOldMangling,
...
     name += kAnalyzerDelimiter;
+
     name += analyzer._shortName;
   } else {
-    name.append(kStirngSuffix);
+    name.append(kStringSuffix);
+    if (analyzer->requireMangled()) {
+      name += kAnalyzerDelimiter;
+      name += analyzer._shortName;
+    }
```
Comment (MBkkt) [inline arangod/IResearch/IResearchKludge.cpp:124]:
```suggestion
```

File: arangod/IResearch/IResearchKludge.cpp [hunk 20]
```diff
@@ -114,10 +114,33 @@ void mangleField(std::string& name, bool isOldMangling,
   normalizeExpansion(name);
   if (isOldMangling || analyzer._pool->requireMangled()) {
     name += kAnalyzerDelimiter;
+
```
Comment (MBkkt) [inline arangod/IResearch/IResearchKludge.cpp:117]:
```suggestion
```

**Review by MBkkt — CHANGES_REQUESTED** [review]
Also please remove velocypack and rocksdb changes

File: arangod/IResearch/SearchDoc.h [hunk 21]
```diff
@@ -0,0 +1,86 @@
...
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Andrey Abramo
```
Comment (Dronplane) [inline arangod/IResearch/SearchDoc.h:21]:
```suggestion
/// @author Andrey Abramov
```
:)

File: arangod/Aql/OptimizerRulesFeature.cpp [hunk 22]
```diff
@@ -483,6 +483,13 @@ void OptimizerRulesFeature::addRules() {
                OptimizerRule::lateDocumentMaterializationRule,
                OptimizerRule::makeFlags(OptimizerRule::Flags::CanBeDisabled));
 
+#ifdef USE_ENTERPRISE
+  // apply late materialization for view queries
```
Comment (Dronplane) [inline arangod/Aql/OptimizerRulesFeature.cpp:487]:
```suggestion
  // handle offsets for ArangoSearch
```

File: arangod/Aql/OptimizerRule.h [hunk 23]
```diff
@@ -341,6 +341,12 @@ struct OptimizerRule {
     // for arango search view
     lateDocumentMaterializationArangoSearchRule,
 
+#ifdef USE_ENTERPRISE
+    // FIXME(gnusi): when to execute?
+    // handle OFFSET_INFO functions calls
+    hanldeOffsetInfoFunc,
```
Comment (Dronplane) [inline arangod/Aql/OptimizerRule.h:347]:
```suggestion
    handleOffsetInfoFunc,
```
typo

**Review by Dronplane — APPROVED** [review]
LGTM. Some minor comments.
And do we really need empty arangod/IResearch/IResearchSearchFunc.cpp?

**Pull request merged**